### ISSUE TITLE
Concurrent GET requests can cause cache misses?

### ISSUE THREAD

**Issue by asajeffrey** [whatwg/fetch#932]
According to https://fetch.spec.whatwg.org/#http-network-or-cache-fetch when doing an http-network-or-cache-fetch, a user agent should first (5.19.1) 

> Set storedResponse to the result of selecting a response from the HTTP cache, possibly needing validation, as per the "Constructing Responses from Caches" chapter of HTTP Caching [HTTP-CACHING], if any.

and later (7.5.2)

> Store httpRequest and forwardResponse in the HTTP cache, as per the "Storing Responses in Caches" chapter of HTTP Caching. [HTTP-CACHING]

As far as I can tell, there's no locking mechanism, so if two GET requests for the same resource arrive concurrently, and the second arrives before the response for the first, then a UA is required to treat it as a cache miss and refetch the data.

This is in line with RFC 7234, https://httpwg.org/specs/rfc7234.html#rfc.section.3.1 which only allows storing partial responses after the status code has arrived, but doesn't seem ideal.

**Comment by asajeffrey** [comment]
The matching Servo issue: https://github.com/servo/servo/issues/24166

**Comment by mnot** [comment]
7234 does not require that two GET requests be separated by a cache. It is *extremely* common for caches (at least in intermediaries) to "collapse" overlapping requests for the same URL by holding any after the first until the response headers are received; if the response is cacheable, it's fanned out to all of the queued requests.

This usually entails having a marker object in cache for those responses that aren't cacheable, so you don't needlessly block requests for things that probably won't benefit from this. Not sure how important that would be in a browser context, though.

For more info:
  * https://wiki.squid-cache.org/Features/CollapsedForwarding
  * https://cwiki.apache.org/confluence/display/TS/Collapsed+Forwarding
  * https://docs.fastly.com/en/guides/request-collapsing

**Comment by asajeffrey** [comment]
Yeah, in practice I'm sure caches do collapse requests, and use locking to ensure atomicity, I just don't see how the fetch spec language allows it? 5.19.1 only reads from the cache, and 7.5.2 stores in the cache after the status code has arrived?
